### PR TITLE
Fix an issue where ENABLE_INTRA_NODE_COMM=1 + multiple process groups leads to failure

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130269

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

**Review by Chillee — APPROVED** [review]

File: torch/csrc/distributed/c10d/intra_node_comm.cpp [hunk 1]
```diff
@@ -31,6 +31,8 @@ static std::vector<std::string> ENABLE_INTRA_NODE_COMM = {
 // for testing purposes.
 static std::vector<std::string> TEST_INTRA_NODE_COMM = {"TEST_INTRA_NODE_COMM"};
 
+static int intraNodeCommIdx = 0;
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/intra_node_comm.cpp:34]: I wonder if there's a way to add a test to guard this?

**Comment by jianc99** [comment]
@yifuwang Hi, now I can create new groups. But when I am using this group to do all reduce, like the operation below,
```python
attn.register_forward_hook(lambda _module, _input, output: funcol.all_reduce(output, "sum", group))
```
There is still an error.

```
W0710 18:29:17.898000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570963 closing signal SIGTERM
W0710 18:29:17.901000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570965 closing signal SIGTERM
W0710 18:29:17.906000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570966 closing signal SIGTERM
W0710 18:29:17.908000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570967 closing signal SIGTERM
W0710 18:29:17.910000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570968 closing signal SIGTERM
W0710 18:29:17.912000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570969 closing signal SIGTERM
W0710 18:29:17.923000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:891] Sending process 3570970 closing signal SIGTERM
W0710 18:29:47.925000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:910] Unable to shutdown process 3570965 via 15, forcefully exiting via 9
E0710 18:29:48.715000 140349446473536 torch/distributed/elastic/multiprocessing/api.py:863] failed (exitcode: -11) local_rank: 1 (pid: 3570964) of binary: /fsx-storygen/beidic/anaconda3/envs/fasthesse/bin/python3.11
Traceback (most recent call last):
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/bin/torchrun", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/lib/python3.11/site-packages/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 355, in wrapper
    return f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/lib/python3.11/site-packages/torch/distributed/run.py", line 919, in main
    run(args)
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/lib/python3.11/site-packages/torch/distributed/run.py", line 910, in run
    elastic_launch(
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 138, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/fsx-storygen/beidic/anaconda3/envs/fasthesse/lib/python3.11/site-packages/torch/distributed/launcher/api.py", line 269, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
```

**Comment by jianc99** [comment]
And this code can successfully run before the introduction of SymmetryMemory, but now it can not run. And if I just use the original nccl allreduce kernel, the code can successfully run as well, can you check that?

**Comment by yifuwang** [comment]
@jianc99 I think this'll fix it https://github.com/pytorch/pytorch/pull/130492

**Pull request closed without merging**